src/main/java/org/sonatype/nexus/ci/nxrm/v3/CreateTagBuilder.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2016-present Sonatype, Inc. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package org.sonatype.nexus.ci.nxrm.v3;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+import com.sonatype.nexus.api.exception.RepositoryManagerException;
+import com.sonatype.nexus.api.repository.v3.RepositoryManagerV3Client;
+import com.sonatype.nexus.api.repository.v3.Tag;
+
+import org.sonatype.nexus.ci.util.NxrmUtil;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import hudson.EnvVars;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.model.AbstractProject;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.Builder;
+import hudson.util.FormValidation;
+import hudson.util.ListBoxModel;
+import jenkins.tasks.SimpleBuildStep;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+
+import static com.sonatype.nexus.api.common.NexusStringUtils.isBlank;
+import static com.sonatype.nexus.api.common.NexusStringUtils.isNotBlank;
+import static hudson.Util.fixEmptyAndTrim;
+import static hudson.model.Result.FAILURE;
+import static hudson.util.FormValidation.error;
+import static hudson.util.FormValidation.ok;
+import static org.sonatype.nexus.ci.config.NxrmVersion.NEXUS_3;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_DisplayName;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Error_TagAttributesJson;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Error_TagAttributesPath;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Validation_TagAttributesJson;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Validation_TagNameEmpty;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Validation_TagNameInvalid;
+import static org.sonatype.nexus.ci.util.RepositoryManagerClientUtil.nexus3Client;
+
+public class CreateTagBuilder
+    extends Builder
```

Comment:
@wrightkris I know you dropped a note on my PR as well but I figured I'd add it here as this is the first one through so we can get concensus.  Do we want to settled in on <action>Builder as the names for these or <action>Step. I went the path of 'step' but do not feel strongly. Seems like it would be good to be consistent.